pageserver/src/page_service.rs

```diff
@@ -348,8 +348,13 @@ impl PageServerHandler {
 
         /* Send a tarball of the latest snapshot on the timeline */
 
-        let req_lsn = lsn.unwrap_or_else(|| timeline.get_last_valid_lsn());
-
+        let req_lsn = match lsn {
+            Some(lsn) => {
+                timeline.wait_lsn(lsn)?;
```

Comment:
I've tried to put `wait_lsn()` in both `list_rels` and `list_nonrels` and immediately got a deadlock:
```
walreceiver
   save_decoded_record(100)
      list_rels(100)
         wait_lsn(100) <--- that is deadlock since we didn't call advance_last_record_lsn() yet
      advance_last_record_lsn(100)
```

Then I've tried to put `assert_ne!(thread::current().name(), Some("WAL receiver thread"));` in  `wait_lsn()` -- and that crashes a lot. And such usage can easily lead to deadlocks.

So now I think better strategy would be:
* make all timeline functions in timeline non-waiting
* assert that arg_lsn <= current_record_lsn in all of them
* call wait_lsn explicetly where it is needed (page_service and basebackup)
* assert that wait_lsn isn't called from WalReceiver

That's probably for another PR, for now I've added wait_lsn to list_nonrels -- it seems to be working well.

Reply:
On a second thought we can't actually assert that arg_lsn <= current_record_lsn in timeline function. Same example as in deadlock wouldn't work. Would that be better if we accept enum {LatestLsn, Lsn(lsn)} to explicitly indicate cases where we don't want to go backward in history and just want to receive latest version?